source/language/types.rst

```diff
@@ -704,12 +704,54 @@ can be accessed, using the following general syntax:
    multiDim[0, 0] = 0.0; // multiDim == {{0.0, 1.2}, {2.1, 2.2}, {3.1, 3.2}}
    multiDim[-1, 1] = 0.0; // multiDim == {{0.0, 1.2}, {2.1, 2.2}, {3.1, 0.0}}
 
-The first argument to the ``array`` type constructor is the base type
+The first argument to the ``array`` declaration is the base type
 of the array. The supported classical types include various sizes of ``bit``,
 ``int``, ``uint``, ``float``, ``complex``, and ``angle``, as well as
 ``bool`` and ``duration``. Note that ``stretch`` is not a valid array
 base type.
 
+Arrays cannot be resized or reshaped. Arrays are statically typed, and cannot
+implicitly convert to or from any other type.
+
+The size of an array is constant and immutable, and is recorded once, at
+declaration time.
+
+Arrays allocate a contiguous block of memory of suitable size and alignment to
+accommodate the storage of its elements.
+
+If the array declaration is direct-initialized via initializer list, its elements
+are initialized to the values provided by the initializer list.  Otherwise, the
+memory allocated is not initialized, and that memory's content is undefined.
+
+Partial initialization via initializer list is allowed - i.e. the initializer
+list is incomplete, and does not provide values for all the elements of the array.
+In this case, the elements that are not initialized via initializer list are not
+initialized, and their values are undefined.
+
+As a special case, a direct initializer list declaration of the form:
+
+ ``array[<type>, <size>] <identifier> = { <X> };``
```

Comment:
I can see value in having syntax for initializing all elements to the same value, but this special case syntax that overlaps with the partial-initialization syntax seems unnecessary. (Especially since I imagine the most common form of partial initialization would be to initialize just the first element of say a list of accumulated deltas, which usage is precluded by this special case).

But maybe its ok since you can think of this as *both* partial initialization (where the unspecified elements happen to be initialized to `X`) *and* as constant initialization of the whole array. Assuming we don't care about the cost of initializing more than was needed maybe it's ok. Otherwise, maybe:
```suggestion
As a special case, a direct initializer list declaration of the form:

 ``array[<type>=<X>, <size>] <identifier>;``
```
might be another option for constant initialization of the whole array?

Reply:
I'll chime in and say that I'm not a fan of supporting both partial initialization and the `array[<type>, <size>] <identifier> = { <X> };` initialization above, for the reason that I think the semantics of partial initialization should be basically monotonic (more data specified the initializer list means more elements get initialized). Or to turn this around: it would come as a surprise to me (if I hadn't read the spec) that `array[int, 5] = {3,2}` has more undefined elements than `array[int, 5] = {3}`.